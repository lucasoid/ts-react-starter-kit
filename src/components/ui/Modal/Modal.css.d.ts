export const modalWrap: string;
export const modalDialog: string;
export const small: string;
export const medium: string;
export const large: string;
export const close: string;
