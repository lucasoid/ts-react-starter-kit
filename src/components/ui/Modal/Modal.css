.modalWrap {
    position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: #efefefaa;
}

.modalDialog {
    max-width: 100%;
    margin: 40px auto;
    padding: 1em;
    background: #fff;
    box-sizing: border-box;
}

.modalDialog.small {
    width: 500px;
}

.modalDialog.medium {
    width: 740px;
}

.modalDialog.large {
    width: 1080px;
}

.close {
    text-align: right;
}
