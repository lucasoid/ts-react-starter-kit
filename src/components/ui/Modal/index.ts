export { Modal, ModalWidth } from './Modal';
export const modalRootId: string = 'modal-root';
