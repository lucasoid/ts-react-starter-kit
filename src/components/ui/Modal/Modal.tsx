import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Button } from '~components/ui/Button';
import { modalRootId } from '~components/ui/Modal';
import * as styles from './Modal.css';

export enum ModalWidth {
    SMALL,
    MEDIUM,
    LARGE,
}

interface IModalProps {
    width: ModalWidth;
    onDismiss: Function;
    children?: any;
}

export const Modal: React.FC<IModalProps> = props => {
    let el = document.createElement('div');

    React.useEffect(() => {
        document.getElementById(modalRootId).appendChild(el);
        return () => document.getElementById(modalRootId).removeChild(el);
    });

    React.useEffect(() => {
        const handler: EventListener = (e: KeyboardEvent) => {
            if (e.key === 'Escape') {
                props.onDismiss();
            }
        };
        window.addEventListener('keyup', handler);
        return () => window.removeEventListener('keyup', handler);
    });

    return ReactDOM.createPortal(
        <div className={styles.modalWrap}>
            <div
                role={'dialog'}
                className={`${styles.modalDialog} ${
                    props.width === ModalWidth.SMALL
                        ? styles.small
                        : props.width === ModalWidth.LARGE
                        ? styles.large
                        : styles.medium
                }`}
            >
                <div className={styles.close}>
                    <Button plaintext={true} onClick={props.onDismiss} title={'Close'}>
                        âœ•
                    </Button>
                </div>
                {props.children}
            </div>
        </div>,
        el,
    );
};
