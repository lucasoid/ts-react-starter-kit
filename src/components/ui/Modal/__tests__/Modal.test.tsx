import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { Modal, ModalWidth } from '../Modal';
import { modalRootId } from '../';
import { Themes } from '~theme';

jest.mock('react', () => {
    const original = jest.requireActual('react');
    return {
        __esModule: true,
        ...original,
        useContext: jest.fn(() => ({ theme: 'light' })),
    };
});

describe('Modal', () => {
    beforeEach(() => {
        let el = document.getElementById(modalRootId);
        if (!el) {
            let node = document.createElement('div');
            node.setAttribute('id', modalRootId);
            document.body.appendChild(node);
        }
    });

    it('renders a modal dialog in various widths', () => {
        const onDismiss = jest.fn();
        let wrapper = shallow(
            <Modal onDismiss={onDismiss} width={ModalWidth.SMALL}>
                Some content
            </Modal>,
        );
        expect(wrapper).toMatchSnapshot();

        wrapper = shallow(
            <Modal onDismiss={onDismiss} width={ModalWidth.MEDIUM}>
                Some content
            </Modal>,
        );
        expect(wrapper).toMatchSnapshot();

        wrapper = shallow(
            <Modal onDismiss={onDismiss} width={ModalWidth.LARGE}>
                Some content
            </Modal>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders a themed modal', () => {
        const onDismiss = jest.fn();
        // @ts-ignore
        React.useContext.mockImplementationOnce(() => ({ theme: Themes.LIGHT }));
        let wrapper = shallow(
            <Modal onDismiss={onDismiss} width={ModalWidth.MEDIUM}>
                Some content
            </Modal>,
        );
        expect(wrapper).toMatchSnapshot();

        // @ts-ignore
        React.useContext.mockImplementationOnce(() => ({ theme: Themes.DARK }));
        wrapper = shallow(
            <Modal onDismiss={onDismiss} width={ModalWidth.MEDIUM}>
                Some content
            </Modal>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('closes a modal dialog with the close button', () => {
        const onDismiss = jest.fn();
        let wrapper = mount(
            <Modal onDismiss={onDismiss} width={ModalWidth.MEDIUM}>
                Some content
            </Modal>,
        );
        wrapper.find('button[title="Close"]').simulate('click');
        expect(onDismiss).toBeCalled();
    });

    it('closes a modal dialog with the escape key', () => {
        const onDismiss = jest.fn();
        let wrapper = mount(
            <Modal onDismiss={onDismiss} width={ModalWidth.MEDIUM}>
                Some content
            </Modal>,
        );
        window.dispatchEvent(new KeyboardEvent('keyup', { key: 'Enter' }));
        expect(onDismiss).not.toBeCalled();
        window.dispatchEvent(new KeyboardEvent('keyup', { key: 'Escape' }));
        expect(onDismiss).toBeCalled();
    });

    it('cleans up on unmount', () => {
        const onDismiss = jest.fn();
        let wrapper = mount(
            <Modal onDismiss={onDismiss} width={ModalWidth.MEDIUM}>
                Some content
            </Modal>,
        );
        let domCleanupSpy = jest.spyOn(Node.prototype, 'removeChild');
        let eventCleanupSpy = jest.spyOn(window, 'removeEventListener');
        wrapper.unmount();
        expect(domCleanupSpy).toBeCalled();
        expect(eventCleanupSpy).toBeCalled();
    });
});
