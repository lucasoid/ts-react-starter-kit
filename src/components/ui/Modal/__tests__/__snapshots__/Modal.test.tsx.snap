// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders a modal dialog in various widths 1`] = `
<Portal
  containerInfo={<div />}
>
  <div
    className="modalWrap light"
  >
    <div
      className="modalDialog small"
      role="dialog"
    >
      <div
        className="close"
      >
        <Button
          onClick={[MockFunction]}
          plaintext={true}
          title="Close"
        >
          ✕
        </Button>
      </div>
      Some content
    </div>
  </div>
</Portal>
`;

exports[`Modal renders a modal dialog in various widths 2`] = `
<Portal
  containerInfo={<div />}
>
  <div
    className="modalWrap light"
  >
    <div
      className="modalDialog medium"
      role="dialog"
    >
      <div
        className="close"
      >
        <Button
          onClick={[MockFunction]}
          plaintext={true}
          title="Close"
        >
          ✕
        </Button>
      </div>
      Some content
    </div>
  </div>
</Portal>
`;

exports[`Modal renders a modal dialog in various widths 3`] = `
<Portal
  containerInfo={<div />}
>
  <div
    className="modalWrap light"
  >
    <div
      className="modalDialog large"
      role="dialog"
    >
      <div
        className="close"
      >
        <Button
          onClick={[MockFunction]}
          plaintext={true}
          title="Close"
        >
          ✕
        </Button>
      </div>
      Some content
    </div>
  </div>
</Portal>
`;

exports[`Modal renders a themed modal 1`] = `
<Portal
  containerInfo={<div />}
>
  <div
    className="modalWrap light"
  >
    <div
      className="modalDialog medium"
      role="dialog"
    >
      <div
        className="close"
      >
        <Button
          onClick={[MockFunction]}
          plaintext={true}
          title="Close"
        >
          ✕
        </Button>
      </div>
      Some content
    </div>
  </div>
</Portal>
`;

exports[`Modal renders a themed modal 2`] = `
<Portal
  containerInfo={<div />}
>
  <div
    className="modalWrap dark"
  >
    <div
      className="modalDialog medium"
      role="dialog"
    >
      <div
        className="close"
      >
        <Button
          onClick={[MockFunction]}
          plaintext={true}
          title="Close"
        >
          ✕
        </Button>
      </div>
      Some content
    </div>
  </div>
</Portal>
`;
